Node.js Package Analyzer may run into ConcurrentModificationException

As can be seen in the conversation of #4089 there is a potential for the Node.js Package Analyzer to run into a ConcurrentModificationException under for now unknown circumstances.

Raising a new ticket as #4089 was about the NullPointer that emerged from it while creating the nestedCauseList for an ExceptionCollection when the ConcurrentModificationException (which has a null-valued message) is the first exception in the ExceptionCollection.

As the ConcurrentModificationException occurs in unknown circumstances we had to split fixing the NPE and further research into the ConcurrentModificationException.

Hello,
Just fall into the ConcurrentModificationException too after update maven plugin from 6.5.3 to 7.3.2 :

```
[WARNING] Analyzing `C:\Git\parent\html\default\package-lock.json` - however, the node_modules directory does not exist. Please run `npm install` prior to running dependency-check
Analyzing `C:\Git\aparent\html\default\package-lock.json` - however, the node_modules directory does not exist. Please run `npm install` prior to running dependency-check
dependency skipped: node module fsevents seems optional and not installed
dependency skipped: node module fsevents seems optional and not installed
dependency skipped: node module fsevents seems optional and not installed
An unexpected error occurred during analysis of 'C:\Git\parent\skin\mod\package-lock.json' (Node.js Package Analyzer): null

[ERROR] 
java.util.ConcurrentModificationException
    at java.util.TreeMap$PrivateEntryIterator.nextEntry (TreeMap.java:1211)
    at java.util.TreeMap$KeyIterator.next (TreeMap.java:1265)
    at java.lang.Iterable.forEach (Iterable.java:74)
    at java.util.Collections$UnmodifiableCollection.forEach (Collections.java:1080)
    at org.owasp.dependencycheck.analyzer.DependencyMergingAnalyzer.mergeDependencies (DependencyMergingAnalyzer.java:156)
    at org.owasp.dependencycheck.analyzer.NodePackageAnalyzer.processDependencies (NodePackageAnalyzer.java:420)
    at org.owasp.dependencycheck.analyzer.NodePackageAnalyzer.processDependencies (NodePackageAnalyzer.java:365)
    at org.owasp.dependencycheck.analyzer.NodePackageAnalyzer.analyzeDependency (NodePackageAnalyzer.java:266)
    at org.owasp.dependencycheck.analyzer.AbstractAnalyzer.analyze (AbstractAnalyzer.java:131)
    at org.owasp.dependencycheck.AnalysisTask.call (AnalysisTask.java:88)
    at org.owasp.dependencycheck.AnalysisTask.call (AnalysisTask.java:37)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
```

After some digging, reverting commit 1bded568d3fcf4bb7d445ba508e96a31a56d7342 seems to solve this issue.
However, I don't run all tests.

NOTE: I was a little bit disappointed since I do not want any NodeJS analysis on this project and have already set those properties :
```
        <nodeAnalyzerEnabled>false</nodeAnalyzerEnabled>
        <pnpmAuditAnalyzerEnabled>false</pnpmAuditAnalyzerEnabled>
        <retireJsAnalyzerEnabled>false</retireJsAnalyzerEnabled>
        <yarnAuditAnalyzerEnabled>false</yarnAuditAnalyzerEnabled>
```
So I can't understand why NodePackageAnalyzer is still launched.

I see this error with Version 7.4.1 too:

`[INFO] Analysis Started
[INFO] Finished Archive Analyzer (2 seconds)
[INFO] Finished File Name Analyzer (0 seconds)
[INFO] Finished Jar Analyzer (1 seconds)
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] An unexpected error occurred during analysis of '/home/jenkins/agent/workspace/xxx/package.json' (Node.js Package Analyzer): null
[ERROR] 
java.util.ConcurrentModificationException
    at java.util.TreeMap$PrivateEntryIterator.nextEntry (TreeMap.java:1208)
    at java.util.TreeMap$KeyIterator.next (TreeMap.java:1262)
    at java.lang.Iterable.forEach (Iterable.java:74)
    at java.util.Collections$UnmodifiableCollection.forEach (Collections.java:1085)
    at org.owasp.dependencycheck.analyzer.DependencyMergingAnalyzer.mergeDependencies (DependencyMergingAnalyzer.java:156)
    at org.owasp.dependencycheck.analyzer.NodePackageAnalyzer.processDependencies (NodePackageAnalyzer.java:456)
    at org.owasp.dependencycheck.analyzer.NodePackageAnalyzer.analyzeDependency (NodePackageAnalyzer.java:270)
    at org.owasp.dependencycheck.analyzer.AbstractAnalyzer.analyze (AbstractAnalyzer.java:131)
    at org.owasp.dependencycheck.AnalysisTask.call (AnalysisTask.java:88)
    at org.owasp.dependencycheck.AnalysisTask.call (AnalysisTask.java:37)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[WARNING] No lock file exists - this will result in false negatives; please run `npm install --package-lock`
[INFO] Finished Node.js Package Analyzer (0 seconds)
[INFO] Finished Dependency Merging Analyzer (0 seconds)
[INFO] Finished Version Filter Analyzer (0 seconds)
[INFO] Finished Hint Analyzer (0 seconds)
[INFO] Created CPE Index (1 seconds)
[INFO] Finished CPE Analyzer (3 seconds)
[INFO] Finished False Positive Analyzer (0 seconds)
[INFO] Finished NVD CVE Analyzer (0 seconds)
[INFO] Finished RetireJS Analyzer (3 seconds)
[INFO] Finished Sonatype OSS Index Analyzer (7 seconds)
[INFO] Finished Vulnerability Suppression Analyzer (0 seconds)
[INFO] Finished Dependency Bundling Analyzer (0 seconds)
[INFO] Finished Unused Suppression Rule Analyzer (0 seconds)
[INFO] Analysis Complete (18 seconds)`

--- a/core/src/main/java/org/owasp/dependencycheck/analyzer/DependencyMergingAnalyzer.java
+++ b/core/src/main/java/org/owasp/dependencycheck/analyzer/DependencyMergingAnalyzer.java
@@ -152,9 +152,11 @@ public class DependencyMergingAnalyzer extends AbstractDependencyComparingAnalyz
             relatedDependency.getEvidence(EvidenceType.PRODUCT).forEach((e) -> dependency.addEvidence(EvidenceType.PRODUCT, e));
             relatedDependency.getEvidence(EvidenceType.VERSION).forEach((e) -> dependency.addEvidence(EvidenceType.VERSION, e));
 
-            relatedDependency.getRelatedDependencies()
-                    .forEach(dependency::addRelatedDependency);
-            relatedDependency.clearRelatedDependencies();
+            synchronized (relatedDependency) {
+                relatedDependency.getRelatedDependencies()
+                        .forEach(dependency::addRelatedDependency);
+                relatedDependency.clearRelatedDependencies();
+            }
             dependency.addAllProjectReferences(relatedDependency.getProjectReferences());
             if (dependenciesToRemove != null) {
                 dependenciesToRemove.add(relatedDependency);

--- a/core/src/main/java/org/owasp/dependencycheck/analyzer/DependencyMergingAnalyzer.java
+++ b/core/src/main/java/org/owasp/dependencycheck/analyzer/DependencyMergingAnalyzer.java
@@ -48,6 +48,10 @@ public class DependencyMergingAnalyzer extends AbstractDependencyComparingAnalyz
      * The phase that this analyzer is intended to run in.
      */
     private static final AnalysisPhase ANALYSIS_PHASE = AnalysisPhase.POST_INFORMATION_COLLECTION;
+    /**
+     * Used for synchronization when merging related dependencies.
+     */
+    private static final Object DEPENDENCY_LOCK = new Object();
 
     /**
      * Returns the name of the analyzer.
@@ -145,18 +149,16 @@ public class DependencyMergingAnalyzer extends AbstractDependencyComparingAnalyz
      */
     public static void mergeDependencies(final Dependency dependency, final Dependency relatedDependency,
             final Set<Dependency> dependenciesToRemove) {
-        synchronized (dependency) {
+        synchronized (DEPENDENCY_LOCK) {
             LOGGER.debug("Merging '{}' into '{}'", relatedDependency.getFilePath(), dependency.getFilePath());
             dependency.addRelatedDependency(relatedDependency);
             relatedDependency.getEvidence(EvidenceType.VENDOR).forEach((e) -> dependency.addEvidence(EvidenceType.VENDOR, e));
             relatedDependency.getEvidence(EvidenceType.PRODUCT).forEach((e) -> dependency.addEvidence(EvidenceType.PRODUCT, e));
             relatedDependency.getEvidence(EvidenceType.VERSION).forEach((e) -> dependency.addEvidence(EvidenceType.VERSION, e));
 
-            synchronized (relatedDependency) {
-                relatedDependency.getRelatedDependencies()
-                        .forEach(dependency::addRelatedDependency);
-                relatedDependency.clearRelatedDependencies();
-            }
+            relatedDependency.getRelatedDependencies()
+                    .forEach(dependency::addRelatedDependency);
+            relatedDependency.clearRelatedDependencies();
             dependency.addAllProjectReferences(relatedDependency.getProjectReferences());
             if (dependenciesToRemove != null) {
                 dependenciesToRemove.add(relatedDependency);

